Fix animation restarts, implement support for non-looped animations

MirceaKitsune opened the pull request:
This fixes an important bug with animated models, and also introduces an important new feature. I tested both changes and found no issues.

Bug: Animations often restart before they finish playing, causing the animation to randomly jump and interrupt. This happens because multiple packages are sometimes received for the same animation, and re-apply an animation that's already ongoing. I've added checks so only animation properties that actually change are applied to the model.

Feature: Ever since animated models were implemented, people requested the ability to have animations play once instead of looping. I added a new parameter to the set_animation function, which can be set to false to disable looping. It of course defaults to true if nothing is set, so behavior won't change for existing mods. This makes proper death and sitting animations possible, and I should have seriously done it a long time ago.

MirceaKitsune commented:
Regarding why the Travis build fails: Apparently it uses an old version of Irrlicht, which doesn't have the setLoopMode function, which I'm using to enable one-time animations. openSUSE uses Irrlicht 1.8.1, which is way over an year old. Therefore I dare assume that it's the fault of Travis CI for using an ancient version... the developers can of course clarify.

ShadowNinja commented on src/content_cao.cpp:
```diff
@@ -1459,9 +1460,15 @@ void GenericCAO::updateAnimation()
-	m_animated_meshnode->setAnimationSpeed(m_animation_speed);
+
+	if (m_animated_meshnode->getStartFrame() != m_animation_range.X ||
+	m_animated_meshnode->getEndFrame() != m_animation_range.Y)
```
This line should be double-tabbed, like
```
if (m_animated_meshnode->getStartFrame() != m_animation_range.X ||
                m_animated_meshnode->getEndFrame() != m_animation_range.Y)
```

MirceaKitsune replied:
Fixed.

ShadowNinja commented on src/content_sao.cpp:
```diff
@@ -366,7 +367,7 @@ std::string LuaEntitySAO::getClientInitializationData(u16 protocol_version)
 		os<<serializeLongString(getPropertyPacket()); // message 1
 		os<<serializeLongString(gob_cmd_update_armor_groups(m_armor_groups)); // 2
-		os<<serializeLongString(gob_cmd_update_animation(m_animation_range, m_animation_speed, m_animation_blend)); // 3
+		os<<serializeLongString(gob_cmd_update_animation(m_animation_range, m_animation_speed, m_animation_blend, m_animation_loop)); // 3
```
This line is way too long, break it into a few lines.

MirceaKitsune replied:
I'm not sure exactly how to break this into multiple lines, nor if I should do it in this patch. The line is actually as long as others above / below it, so I'd need to break all of them. And if I started doing this, I'd fill the patch with a lot of code style cleanups rather than focusing on the feature. I'd suggest someone else does this separately afterward, for all lines which are considered too long... although they seem fine to me personally.

ShadowNinja commented on src/content_cao.cpp:
```diff
@@ -1644,6 +1653,8 @@ void GenericCAO::processMessage(const std::string &data)
 				m_animation_speed = readF1000(is);
 				m_animation_blend = readF1000(is);
+				// these are sent inverted so we get true when the server sends nothing
+				m_animation_loop = !readU8(is);
```
I'd prefer if:
  * This was a more generic flags field.
  * This wasn't inverted, but used a length check instead.

MirceaKitsune replied:
I'm not sure how to do these. I just used what other functions use to network booleans over. If something should be changed, it should probably be changed for all functions in a row, perhaps as a different patch later on.

Zeno- replied:
Or, instead of inverting what's serialised "invert" the variable name; e.g. m_animation_loop => m_animation_once (or play once, or do not loop, or whatever) ;)

ShadowNinja commented:
Minetest currently supports 1.7, and I don't think we're going to drop it because of this change.  If 1.7 can't support this it's better to have a graphical glitch than a compile error.

MirceaKitsune commented:
@ShadowNinja Sorry, but that I strongly disagree with. Why would anyone prefer leaving a bug in the code or not implementing a long awaited feature, only to maintain compatibility with an ancient Irrlicht version? 1.8.1 has been out since 2013, and there's no reason why any developer or Linux distribution wouldn't have it by now. I'd like other developers to please clarify me on this.

MirceaKitsune commented:
The Irrlicht 1.7 compatibility issue was solved with #if tags. Everyone will get the fix for the animation restart bug, but non-looped animations will only work for people with Irrlicht 1.8 or greater, while the rest will still see them loop.

nerzhul commented:
Code seems good for me. Except the inversion pointed by ShadowNinja

MirceaKitsune commented:
@nerzhul I followed the model used by other booleans. More specifically, what gob_cmd_update_physics_override does with sneak and sneak_glitch. If it's wrong, it should probably be changed for these as well. I don't personally know of a better way... so let me know exactly how to write them, or a better coder can change it after or before this is pulled.

nerzhul commented:
Warn, incompatible change here, this modify the datas passed over network. I add the protocol bump tag. This PR is reserved to dev-0.5 branch.

MirceaKitsune commented:
Is this certain? I've only added a new parameter to the existing function. I'm assuming that old clients would simply ignore it and work as before, so this shouldn't break any compatibility.

In either case I can wait. My only concern would be rebase conflicts, since if anyone adds another function near this I'll have a few dozen lines of code to fix :)

nerzhul commented:
Yes but your parameter is sent into serialized data sent to client.

MirceaKitsune commented:
I know, it is technically a different protocol architecture. Still, old clients should just ignore the extra setting, so compatibility would normally be fine. I thought protocol bumps happened when entirely new functions were implemented and networked over, but I could be wrong.

MirceaKitsune commented:
In any case, I just read that there's a separate branch dev-0.5. Can it be merged in that? I can make a new pull request, but thought to ask first.

nerzhul commented:
No you don't need , if your pr will be accepted we will merge it to.

MirceaKitsune commented:
Oh... good in that case. This is still ongoing for the dev-0.5 branch then.

nerzhul commented:
@Zeno- @ShadowNinja this commit is okay for me into dev-0.5, are you okay  too ?

ShadowNinja commented:
@nerzhul This doesn't break anything (or at least shouldn't and doesn't need to), so it should be merged into master, dev-0.5 is just for breaking changes as I understand it.

nerzhul commented:
Then we could merge it in master.

MirceaKitsune commented:
Like I said, I tested both the animation twitch fix and non-loop feature as well. Both seem to work correctly and I saw no problems.

MirceaKitsune commented:
I'd like to know if there is any news on this please. Can this fix be merged into the development branch as discussed above?

MirceaKitsune commented:
Rebased and fixed merge conflicts, caused by the addition of the get_animation function. The new flag was added to get_animation in the process, which will report it like all other properties.

est31 commented on src/content_cao.cpp:
```diff
@@ -1465,9 +1466,18 @@ void GenericCAO::updateAnimation()
-	m_animated_meshnode->setFrameLoop(m_animation_range.X, m_animation_range.Y);
-	m_animated_meshnode->setAnimationSpeed(m_animation_speed);
+
+	if (m_animated_meshnode->getStartFrame() != m_animation_range.X ||
```
Why first getting before setting? Is there a specific reason?

MirceaKitsune replied:
Yes: It stops the animation from restarting if multiple animation packages are received at once by the client. Basically it doesn't re-set the animation if the parameters are exactly the same, which has been an annoyance I didn't notice and fix when I initially implemented the code.

est31 commented:
We have [these](http://dev.minetest.net/Code_style_guidelines) formatting guidelines. Much of the code has been written before they got established, so it doesn't abide them. New code and code changes should follow them.

est31 commented:
will look at it more thoroughly the next days.

MirceaKitsune commented:
@est31 Ok. The problem that confused me is that I've implemented a new parameter and a bugfix, but instead of the changes being discussed I only hear about lines being too long. I only did the exact same thing as the lines above and before mine (which are even longer), especially since I didn't want the code looking disproportional so I aimed at formatting everything the same way.

My opinion remains that if the lines are too long, they should be fixed in a separate commit for ALL functions. That is, after this change is merged... I don't wish to solve more merge conflicts, please. For now, I broke apart the lines I was asked to in my commit, so I understand it should be good.

est31 commented:
:+1: looks good.

About the protocol change:
As we explicitly know the size of the active object messages, your way to do it is IMO perfectly fine. Perhaps add a "//added with protocol 25" comment, so that one knows when to break backwards compatibility in the future.

est31 commented:
660fa516bfe774c77947c47a97154d6f069f414d
